src/code/audio_thread_manager.c

```diff
@@ -1,91 +1,150 @@
         Sched_Notify(audioMgr->sched);
     }
 
+    // Update the audio driver
+
     gAudioThreadUpdateTimeStart = osGetTime();
-    if (SREG(20) >= 2) {
+
+    if (R_AUDIOMGR_DEBUG_LEVEL >= AUDIOMGR_DEBUG_LEVEL_NO_UPDATE) {
+        // Skip update, no rsp task produced
         rspTask = NULL;
     } else {
         rspTask = func_800E4FE0();
     }
+
     gAudioThreadUpdateTimeAcc += osGetTime() - gAudioThreadUpdateTimeStart;
     gAudioThreadUpdateTimeStart = 0;
 
     if (audioMgr->rspTask != NULL) {
-        osRecvMesg(&audioMgr->taskQueue, NULL, OS_MESG_BLOCK);
-        func_800C3C80(audioMgr);
+        // Wait for the audio rsp task scheduled on the previous retrace to complete. This looks like it should wait
+        // for the task scheduled on the current retrace, but since the queue is initially filled in AudioMgr_Init
+        // it doesn't wait for the task scheduler to post a message for the most recent task as there is already a
+        // message waiting.
+        osRecvMesg(&audioMgr->taskDoneQueue, NULL, OS_MESG_BLOCK);
+        // Report task done
+        //! @bug As a result of above, the task done notification is sent to the next task instead of the task it was
+        //! intended for. In practice, task done notifications are not used by the audio driver so this is
+        //! inconsequential.
+        AudioMgr_NotifyTaskDone(audioMgr);
     }
+    // Update rsp task to be scheduled on next retrace
     audioMgr->rspTask = rspTask;
 }
 
+/**
+ * Handle Pre-NMI event.
+ * Implemented by the audio driver.
+ *
+ * @see Audio_PreNMI
+ */
 void AudioMgr_HandlePreNMI(AudioMgr* audioMgr) {
     // "Audio manager received OS_SC_PRE_NMI_MSG"
     osSyncPrintf("オーディオマネージャが OS_SC_PRE_NMI_MSG を受け取りました\n");
     Audio_PreNMI();
 }
 
-void AudioMgr_ThreadEntry(void* arg0) {
-    AudioMgr* audioMgr = (AudioMgr*)arg0;
+void AudioMgr_ThreadEntry(void* arg) {
+    AudioMgr* audioMgr = (AudioMgr*)arg;
     IrqMgrClient irqClient;
     s16* msg = NULL;
 
-    osSyncPrintf("オーディオマネージャスレッド実行開始\n"); // "Start running audio manager thread"
+    // "Start running audio manager thread"
+    osSyncPrintf("オーディオマネージャスレッド実行開始\n");
+
+    // Initialize audio driver
     Audio_Init();
     AudioLoad_SetDmaHandler(DmaMgr_AudioDmaHandler);
     Audio_InitSound();
-    osSendMesg(&audioMgr->lockQueue, NULL, OS_MESG_BLOCK);
+
+    // Fill init queue to signal that the audio driver is initialized
+    osSendMesg(&audioMgr->initQueue, NULL, OS_MESG_BLOCK);
+
     IrqMgr_AddClient(audioMgr->irqMgr, &irqClient, &audioMgr->interruptQueue);
 
+    // Spin waiting for events
     while (true) {
         osRecvMesg(&audioMgr->interruptQueue, (OSMesg*)&msg, OS_MESG_BLOCK);
+
         switch (*msg) {
             case OS_SC_RETRACE_MSG:
                 AudioMgr_HandleRetrace(audioMgr);
+
+                // Empty the interrupt queue
                 while (!MQ_IS_EMPTY(&audioMgr->interruptQueue)) {
```

Comment:
Is there a reason to have another inner while loop instead of just having the one outer loop?

Reply:
The idea is to only handle a retrace event once until the queue is emptied, I guess if the retrace handling takes so long that another retrace event is already waiting they didn't want to run it again so quickly

Reply:
If that's ever the case you are having big problems anyway, but I guess this is still better than monopolizing the cpu on audio then :shrug: